### PR TITLE
Extend ltinyiiod

### PR DESCRIPTION
**Pull request by cristiantpop**
Extend libtinyiiod interface to support serial and network interface.

### CODE REVIEW

File: tinyiiod.c [hunk 1]
```diff
@@ -96,7 +93,7 @@ ssize_t tinyiiod_read_line(struct tinyiiod *iiod, char *buf, size_t len)
 
 	buf[i] = '\0';
 
-	return (ssize_t) i;
+	return i;
```
Comment (stefpopa) [inline tinyiiod.c:99]: hmm.. is this intended? This function returns `uint32_t `instead of `int32_t` as it would be expected from the function type.

File: tinyiiod.c [hunk 2]
```diff
@@ -46,12 +43,12 @@ void tinyiiod_destroy(struct tinyiiod *iiod)
...
 
-int tinyiiod_read_command(struct tinyiiod *iiod)
+int32_t tinyiiod_read_command(struct tinyiiod *iiod)
 {
 	char buf[128];
-	int ret;
+	int32_t ret;
 
-	ret = (int) tinyiiod_read_line(iiod, buf, sizeof(buf));
+	ret = tinyiiod_read_line(iiod, buf, sizeof(buf));
```
Comment (stefpopa) [inline tinyiiod.c:51]: Currently, this function returns `uint32_t`, so `ret `will always be greater than 0

**Comment by stefpopa** [comment]
As a general remark, the commit messages should not exceed 74 characters or so. From this perspective, your commit messages are fine, however you probably can configure your editor to automatically break after 74 columns, so won't have to do it manually.

File: tinyiiod.c (hunk 1 above)
Comment (cristiantpop) [reply tinyiiod.c:99]: I will change 
uint32_t i; 
to 
int32_t i;

File: CMakeLists.txt [hunk 3]
```diff
@@ -44,6 +44,7 @@ if (NOT TINYIIOD_VERSION_GIT)
 	set(TINYIIOD_VERSION_GIT v${TINYIIOD_VERSION})
 endif()
 
+#add_definitions(-D_COMPAT_STDINT_TYPES)
```
Comment (mirceacaprioru) [inline CMakeLists.txt:47]: nitpick : you have an extra e in free.

**Comment by stefpopa** [comment]
For better traceability of what was fixed, please add a change log.

File: tinyiiod.c [hunk 4]
```diff
@@ -183,37 +174,68 @@ void tinyiiod_do_close(struct tinyiiod *iiod, const char *device)
...
+			     const char *device, size_t bytes_count)
+{
+	int32_t ret;
+	char *pbuffer = (char*)malloc(bytes_count);
+	if(pbuffer == NULL) {
+		ret = -ENOMEM;
+		goto err_close;
+	}
+	tinyiiod_read_nonblocking(iiod, pbuffer, bytes_count);
+	tinyiiod_write_value(iiod, (int) bytes_count);
```
Comment (stefpopa) [inline tinyiiod.c:216]: Should we check the return of the above functions in case of error?

File: tinyiiod-private.h [hunk 5]
```diff
@@ -41,9 +41,13 @@ void tinyiiod_do_write_attr(struct tinyiiod *iiod, const char *device,
...
 void tinyiiod_do_close(struct tinyiiod *iiod, const char *device);
+int32_t tinyiiod_do_close_instance(struct tinyiiod *iiod);
 
-void tinyiiod_do_readbuf(struct tinyiiod *iiod,
-			 const char *device, size_t bytes_count);
+int32_t tinyiiod_do_readbuf(struct tinyiiod *iiod,
+			    const char *device, size_t bytes_count);
+
+int32_t tinyiiod_do_writebuf(struct tinyiiod *iiod, const char *device,
+			     size_t bytes_count);
```
Comment (stefpopa) [inline tinyiiod-private.h:50]: From my understanding, this commit implements some new functions, such as` tinyiiod_do_writebuf() ` and modifies the implementation of already existing functions, such as `tinyiiod_do_readbuf()`. This is fine.
However, it looks like it also tries to change the return type of some functions, such as `tinyiiod_read()`. I would add these changes in a different commit. It makes the patch cleaner and easier to follow.

File: tinyiiod.c [hunk 6]
```diff
@@ -183,37 +174,68 @@ void tinyiiod_do_close(struct tinyiiod *iiod, const char *device)
...
+	size_t bytes_received = tinyiiod_read_wait(iiod, bytes_count);
+	if(bytes_count != bytes_received) {
+		ret = -EPIPE;
+		goto err_close;
+	}
+
+	iiod->ops->write_device(device, pbuffer, bytes_count);
+	tinyiiod_write_value(iiod, (int) bytes_count);
```
Comment (stefpopa) [inline tinyiiod.c:226]: same question as [above](https://github.com/analogdevicesinc/libtinyiiod/pull/6/commits/c55bc8e96b9a465e60335a3b455632442e92011b#r259260017)

File: example.c [hunk 7]
```diff
@@ -21,9 +21,9 @@
 #include <stdio.h>
 #include <string.h>
 
-static void read_data(char *buf, size_t len)
```
Comment (mirceacaprioru) [inline example.c:24]: In my opinion it seems this commit is actually 3-4 commits put together. I would separate different actions into their own commit.

File: tinyiiod.c [hunk 8]
```diff
@@ -67,33 +67,24 @@ char tinyiiod_read_char(struct tinyiiod *iiod)
...
 }
 
-void tinyiiod_read(struct tinyiiod *iiod, char *buf, size_t len)
+int32_t tinyiiod_read(struct tinyiiod *iiod, char *buf, size_t len)
 {
-	iiod->ops->read(buf, len);
+	return iiod->ops->read(buf, len);
 }
 
 int32_t tinyiiod_read_line(struct tinyiiod *iiod, char *buf, size_t len)
```
Comment (mirceacaprioru) [inline tinyiiod.c:75]: do we need this wrapper anymore ?

File: tinyiiod.c [hunk 9]
```diff
@@ -183,37 +174,68 @@ void tinyiiod_do_close(struct tinyiiod *iiod, const char *device)
...
+{
+	return iiod->ops->close_instance();
+}
+
+int32_t tinyiiod_do_writebuf(struct tinyiiod *iiod,
+			     const char *device, size_t bytes_count)
+{
+	int32_t ret;
+	char *pbuffer = (char*)malloc(bytes_count);
+	if(pbuffer == NULL) {
```
Comment (stefpopa) [inline tinyiiod.c]: Nitpick of the day: `if (!pbuffer)`

File: tinyiiod.c (hunk 8 above)
Comment (cristiantpop) [reply tinyiiod.c:75]: this will be removed in a later pull request.

Comment (mirceacaprioru) [reply tinyiiod.c:75]: ack

**Comment by cristiantpop** [comment]
changelog v1 -> v2

* fixed upper problems

File: compat.h [hunk 10]
```diff
@@ -0,0 +1,56 @@
+/*
+ * libtinyiiod - Tiny IIO Daemon Library
+ *
+ * Copyright (C) 2016 Analog Devices, Inc.
+ * Author: Paul Cercueil <paul.cercueil@analog.com>
```
Comment (commodo) [inline compat.h]: you can put yourself as author here and update the year to 2019
and change `libtinyiiod - Tiny IIO Daemon Library`  ->  `libc compatibility layer` or something like that

File: tinyiiod.c [hunk 11]
```diff
@@ -164,6 +174,40 @@ void tinyiiod_do_close(struct tinyiiod *iiod, const char *device)
 	tinyiiod_write_value(iiod, ret);
 }
 
+int32_t tinyiiod_do_close_instance(struct tinyiiod *iiod)
```
Comment (commodo) [inline tinyiiod.c:199]: how does `close_instance` differ from `close()` ?

i saw the comment: `tinyiiod_do_close_instance() - Multiple clients can connect simultaneously at tinyiiod, trough libtinyiiod.`  but there is no `open_instance()` or similar behavior anywhere

there isn't any instance ID to identify which instance to close either

File: tinyiiod.c [hunk 12]
```diff
@@ -164,6 +174,40 @@ void tinyiiod_do_close(struct tinyiiod *iiod, const char *device)
...
+{
+	return iiod->ops->close_instance();
+}
+
+int32_t tinyiiod_do_writebuf(struct tinyiiod *iiod,
```
Comment (commodo) [inline tinyiiod.c:204]: this doesn't seem to differ much from `tinyiiod_do_write_attr()` in terms of behavior;

if `tinyiiod_read_nonblocking()` is non-blocking, then you could also just get a "no-data-yet" type of error and things fail/error out;

if you do need to implement a `tinyiiod_read_wait()` then you don't need to make this nonblocking

if you extend the buffer in `tinyiiod_do_write_attr()` you pretty much make it identical with this function

File: tinyiiod.h [hunk 13]
```diff
@@ -26,9 +26,12 @@ struct tinyiiod_ops {
 	/* Read from the input stream */
 	void (*read)(char *buf, size_t len);
 	int32_t (*read_line)(char *buf, size_t len);
+	int32_t (*read_nonbloking)(char *buf, size_t len);
```
Comment (commodo) [inline tinyiiod.h]: typo: read_nonbloking -> read_nonblocking

File: tinyiiod.h [hunk 14]
```diff
@@ -44,7 +47,9 @@ struct tinyiiod_ops {
 	int32_t (*open)(const char *device, size_t sample_size, uint32_t mask);
 	int32_t (*close)(const char *device);
 
-	ssize_t (*read_data)(const char *device, char *buf, size_t bytes_count);
+	ssize_t (*read_data)(const char *device, char *pbuf, size_t bytes_count);
```
Comment (commodo) [inline tinyiiod.h]: the `buf` -> `pbuf` rename is just noise; it doesn't need to change

File: tinyiiod.h [hunk 15]
```diff
@@ -44,7 +47,9 @@ struct tinyiiod_ops {
...
+	ssize_t (*write_data)(const char *device, const char *buf,
+			      size_t bytes_count);
```
Comment (commodo) [inline tinyiiod.h:52]: the `const char *` conversion is nice; it breaks the API a bit, but it's not too bad as people would get compiler warnings
still, i'd put it it's own commit

File: tinyiiod.c [hunk 16]
```diff
@@ -77,6 +77,16 @@ int32_t tinyiiod_read_line(struct tinyiiod *iiod, char *buf, size_t len)
 	return iiod->ops->read_line(buf, len);
 }
 
+int32_t tinyiiod_read_nonblocking(struct tinyiiod *iiod, char *buf, size_t len)
```
Comment (commodo) [inline tinyiiod.c:102]: are `tinyiiod_read_nonblocking()` & `tinyiiod_read_wait()` used only in `tinyiiod_do_writebuf()` ?
if so, then maybe we should defer adding them, until we find a better use-case for them

File: tinyiiod.c (hunk 8 above)
Comment (commodo) [reply tinyiiod.c:75]: this `tinyiiod_read_line()` conversion is a bit big
and it should be it's own commit;

though, i wouldn't remove it;

i would do:
```
if (iiod->ops->read_line)
    return iiod->ops->read_line(buf, len);
```

and let the other code remain;
if users want to use their own read_line callback, let's let them; if not, we can fallback to the old one

**Comment by stefpopa** [comment]
> changelog v1 -> v2
> 
> * fixed upper problems

I apologize that I haven't explained in enough detail what the changelog should include and the purpose of it. 
Usually, in the changelog, we try to enumerate whatever was fixed from the previous version. Sometimes when going thorough this step, I realize that I missed one or more comments. Also, it is a sort of acknowledgement from the part of the developer, that the review remarks were taken into consideration and treated accordingly. 
Moreover, by adding a changelog, the reviewers (especially the ones with short memory, like myself and the ones that are too lazy to check on their own review comments, like myself) will be grateful. 
Also, adding a changelog is part of the whole process to get your patches accepted upstream. I know that this is a no-os fix, however, we agreed to follow this process anyway.
A changelog doesn't have to go into much detail. Just a few brief bullet points will do. Here is an example: https://github.com/analogdevicesinc/linux/pull/286#issuecomment-453181191.

File: tinyiiod.h [hunk 17]
```diff
@@ -24,7 +24,7 @@ struct tinyiiod;
 
 struct tinyiiod_ops {
 	/* Read from the input stream */
-	void (*read)(char *buf, size_t len);
+	int32_t (*read)(char *buf, size_t len);
```
Comment (commodo) [inline tinyiiod.h:27]: this change `libtinyiiod: Modify return value for consistency ` is nice;
it shouldn't break API too bad; it would cause compiler warnings though, but we can live with that;

i would change the name to `libtinyiiod: add return value to read() callback` so that the title better reflects what it's doing

File: tinyiiod.c (hunk 11 above)
Comment (cristiantpop) [reply tinyiiod.c:199]: close() and open() are used for closing and opening a device(like cf-ad9361-dds-core-lpc or ad9361-phy). tinyiiod_do_close_instance() to close a network client, with an "EXIT" command received from iio-oscilloscope(libiio). It doesn't have a pair "OPEN" command that comes form iio-osc, so no pair function. Instance is created at a lower level in network.c, network_accept().

File: tinyiiod.c (hunk 12 above)
Comment (cristiantpop) [reply tinyiiod.c:204]: tinyiiod_do_writebuf(), has a tinyiiod_write_value(iiod, (int) bytes_count);, which is a kind of "ack", after this point, client sends data. If the server is not prepared, for receiving of the data (which tinyiiod_read_nonblocking() call does), some bytes are lost. This solution was adopted by me, after a lot of iterations. You can try for better results, until then, I will suggest this one.

File: tinyiiod.h (hunk 13 above)
Comment (cristiantpop) [reply tinyiiod.h]: I will fix this

File: tinyiiod.h (hunk 14 above)
Comment (cristiantpop) [reply tinyiiod.h]: I will fix this

File: tinyiiod.h (hunk 15 above)
Comment (cristiantpop) [reply tinyiiod.h:52]: write_data is something new here. It newer existed before. Please be more explicit.

File: tinyiiod.c (hunk 16 above)
Comment (cristiantpop) [reply tinyiiod.c:102]: They are only used here. tinyiiod_read_nonblocking() prepares the serial interface for a big block of data, since there is no handshake here, like on network interface. tinyiiod_read_wait() waits for entire block of data.

File: tinyiiod.c (hunk 8 above)
Comment (cristiantpop) [reply tinyiiod.c:75]: This is a good idea. If the user doesn't provide it's own implementation we can use this one, that is not so efficient, but it is something.

File: tinyiiod.h (hunk 15 above)
Comment (commodo) [reply tinyiiod.h:52]: my bad here;
i thought it was modifying existing stuff

File: tinyiiod.c (hunk 8 above)
Comment (commodo) [reply tinyiiod.c:75]: looking at how serial_read_line() & network_read_line() are implemented, the  `serial_read()` & `network_read()` callbacks could be used as well;

so, i am un-decided whether to expose `read_line` as an ops

Comment (cristiantpop) [reply tinyiiod.c:75]: This will cause only problems for the layer that uses only serial_read(), since will need to read characters one by one, and check for '\n' or to read chunks of data search for '\n' that will result in even more complications. It is possible to find a '\n' in the middle of the buffer, or maybe you won't find any. So, I would use read_line.

**Comment by cristiantpop** [comment]
changes in v3:

-  put my self as author here and update the year to 2019 in compat.h

- typo: read_nonbloking -> read_nonblocking 

- rename pbuf -> buf; it doesn't need to change

- tinyiiod_read_line() keep the old implementation. Leave user to chose the between, old implementation and the one that he offers.

- reword comment to: libtinyiiod: add return value to read() callback
